File: programmers/무인도여행.js
function solution(maps) {
  var answer = [];
  maps = maps.map((m) => m.split(""));
  //   console.log(maps);

  function dfs(r, c) {
    if (
      r < 0 ||
      c < 0 ||
      r >= maps.length ||
      c >= maps[0].length ||
      maps[r][c] === "X"
    ) {
      return 0;
    } else {
      var num = parseInt(maps[r][c]);
      maps[r][c] = "X";
      return (
        num + dfs(r - 1, c) + dfs(r + 1, c) + dfs(r, c - 1) + dfs(r, c + 1)
      );
    }
  }

  for (let row = 0; row < maps.length; row++) {
    for (let col = 0; col < maps[0].length; col++) {
      if (maps[row][col] !== "X") {
        answer.push(dfs(row, col));
      }
    }
  }

  if (answer.length === 0) {
    return [-1];
  } else {
    console.log(answer.sort((a, b) => a - b));
    return answer.sort((a, b) => a - b);
  }
}

solution(["X591X", "X1X5X", "X231X", "1XXX1"]);

File: programmers/무인도여행.py
def solution(maps:list[str]) -> list[int]:
    answer = []
    maps = list(map(list, maps))

    def dfs(r, c):
        num = 0
        if (r < 0 or c < 0 or r >= len(maps) or c >= len(maps[0]) or maps[r][c] == "X"):
            return 0
        else:
            num += int(maps[r][c])
            maps[r][c] = "X"
            return num + dfs(r-1, c) + dfs(r+1, c) + dfs(r, c-1) + dfs(r, c+1)
    
    for row in range(len(maps)):
        for col in range(len(maps[0])):
            if maps[row][col] != "X":
                answer.append(dfs(row, col))

    if (len(answer) == 0):
        return [-1]
    
    print(sorted(answer))
    return sorted(answer)

solution(["X591X", "X1X5X", "X231X", "1XXX1"])